src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c

```diff
@@ -0,0 +1,458 @@
+    ret = PopulateString(env, loc[algorithm], buf, len);
+
+cleanup:
+    RELEASE_LOCALS(loc, env)
+    return ret;
+}
+
+int32_t AndroidCryptoNative_X509GetPublicKeyBytes(jobject /*X509Certificate*/ cert, uint8_t *buf, int32_t len)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
+
+    int32_t ret = BUFFER_FAIL;
+    INIT_LOCALS(loc, key, keyInfoBytes)
+
+    // PublicKey key = cert.getPublicKey();
+    // byte[] keyInfoBytes = key.getEncoded();
+    loc[key] = (*env)->CallObjectMethod(env, cert, g_X509CertGetPublicKey);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    loc[keyInfoBytes] = (*env)->CallObjectMethod(env, loc[key], g_KeyGetEncoded);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+    // return keyInfoBytes.length;
+    ret = PopulateByteArray(env, loc[keyInfoBytes], buf, len);
+
+cleanup:
+    RELEASE_LOCALS(loc, env)
+    return ret;
+}
+
+int32_t AndroidCryptoNative_X509GetPublicKeyParameterBytes(jobject /*X509Certificate*/ cert, uint8_t *pBuf, int32_t cBuf)
+{
+    // [TODO]
+    // PublicKey key = cert.publicKey()
+    // String algorithm = key.getAlgorithm()
+    // if (algorithm == "...") ...
+    // getParams()
+    return BUFFER_FAIL;
+}
+
+// Serial number as a byte array in big-endian byte-order
+int32_t AndroidCryptoNative_X509GetSerialNumber(jobject /*X509Certificate*/ cert, uint8_t *buf, int32_t len)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
+
+    int32_t ret = BUFFER_FAIL;
+    INIT_LOCALS(loc, serial, bytes)
+
+    // BigInteger serial = cert.getSerialNumber();
+    // buf = serial.toByteArray();
+    // return buf.length;
+    loc[serial] = (*env)->CallObjectMethod(env, cert, g_X509CertGetSerialNumber);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    loc[bytes] = (*env)->CallObjectMethod(env, loc[serial], g_toByteArrayMethod);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    ret = PopulateByteArray(env, loc[bytes], buf, len);
+
+cleanup:
+    RELEASE_LOCALS(loc, env)
+    return ret;
+}
+
+int32_t AndroidCryptoNative_X509GetSignatureAlgorithm(jobject /*X509Certificate*/ cert, char *buf, int32_t len)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
+    int32_t ret = BUFFER_FAIL;
+
+    // String oid = cert.getSigAlgOID()
+    // return oid.length;
+    jstring oid = (jstring)(*env)->CallObjectMethod(env, cert, g_X509CertGetSigAlgOID);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    ret = PopulateString(env, oid, buf, len);
+
+cleanup:
+    (*env)->DeleteLocalRef(env, oid);
+    return ret;
+}
+
+int32_t AndroidCryptoNative_X509GetThumbprint(jobject /*X509Certificate*/ cert, uint8_t *buf, int32_t len)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
+
+    int32_t ret = BUFFER_FAIL;
+    INIT_LOCALS(loc, algorithm, md, encoded, thumbprint)
+
+    // MessageDigest md = MessageDigest.getInstance("SHA-1");
+    loc[algorithm] = JSTRING("SHA-1");
+    loc[md] = (*env)->CallStaticObjectMethod(env, g_mdClass, g_mdGetInstanceMethod, loc[algorithm]);
```

Comment:
If the Java API doesn't have this as precomputed you could just skip the native function and do it in managed code as `SHA1.HashData(cert.RawData)` (or the PAL-non-redundant-allocation version thereof).